### ISSUE TITLE
Mask and Embeddings cause problems with some optimisers.

### ISSUE THREAD

**Issue by dbonadiman** [fchollet/keras#1062]
This has been pointed out different times in different threads. By default Theano at each iteration updates weights for all the weights in a shared matrix. That it is not a problem in the general case.

But there is a case in which this cause problem:

1) take as example for instance the Embedding layer. It is a Huge matrix of weight are filtered according to the input resulting in the case of words in the sentence representation (as a matrix).

2) Use a second order optimisation method (`Adagrad`, `Adadelta`, `RMSprop`) where updates are calculated accordingly to the history of the updates.

From an optimisation point of view a `0` update on a weight means that the weight do not contribute to the final loss. But there are rare words that most of the time are not updated so from the   `accumulators` we obtain a series of 0 updates for this word using this information to adjust the embedding of rare words lead to a near to zero weight update.  The fact that a rare words do not appear in a particular instance means that the some of the weight do not contribute to the final loss because they are not connected and not because they are ok.

This problem  occurs in many cases where the upper layers uses only part of the output of the previous one.  Masked output, Max_pooling and Embeddings are some that for me suffer of that problem.

What do you think? Shouldn't we pass the masks to the optimisers and by doing so updating the weight using the `set_subtensor` method?

To Highlight this issue try to use on the examples using Word Embeddings to use first order optimisers like Adam or SGD, you will notice a much faster convergence. In other cases when you set the word embedding weights (from word2vec for instance) and then you refine them on the specific problem using second order methods you will obtain results near to the one obtained if we set the Embedding layer as not trainable. Differently with Adam you tend to obtain better result.

What do you think guys i'm understanding in the wrong way? or there is an actual problem?

**Comment by dbonadiman** [comment]
There is a page on the theano documentation that explain the problem:
http://deeplearning.net/software/theano/tutorial/faq_tutorial.html

If you are lazy:
>If you want to update only a subset of a weight matrix (such as some rows or some columns) that are used in the forward propogation of each iteration, then the cost function should be defined in a way that it only depends on the subset of weights that are used in that iteration.

And in particular

>Note that you can compute the gradient of the cost function w.r.t. the entire lookup_table, and the gradient will have nonzero rows only for the rows that were selected during forward propagation. If you use gradient descent to update the parameters, there are no issues except for unnecessary computation, e.g. you will update the lookup table parameters with many zero gradient rows. However, if you want to use a different optimization method like rmsprop or Hessian-Free optimization, then there will be issues. In rmsprop, you keep an exponentially decaying squared gradient by whose square root you divide the current gradient to rescale the update step component-wise. If the gradient of the lookup table row which corresponds to a rare word is very often zero, the squared gradient history will tend to zero for that row because the history of that row decays towards zero. Using Hessian-Free, you will get many zero rows and columns. Even one of them would make it non-invertible. In general, it would be better to compute the gradient only w.r.t. to those lookup table rows or columns which are actually used during the forward propagation.

**Comment by EderSantana** [comment]
whoa, important catch!

But Adam also uses accumulators, doesn't it. Is there a way to manually define the gradient os those layers or something like that? Did you see any solution already?

**Comment by dbonadiman** [comment]
Unfortunately i haven't found any solution yet, or better we should calculate the gradient wrt sub = W[X] instead of W. First and then updating the weight on W passing the pair (W, T.set_subtensor(sub, update rule on sub)) Instead of (W, update rule on W).

But is difficult to do in our framework.
BTW, If i remember correctly Adam is a first order optimisation method.

**Comment by EderSantana** [comment]
> BTW, If i remember correctly Adam is a first order optimisation method.

They (rmsprop, adam, adadelta) are all approximately second order (only estimate the diagonal of the Hessian though). I think you can choose hyperparameters in Adam to make it work just like RMSprop for instance.



> But is difficult to do in our framework.

True, we would have to hack up a solution using callbacks and doing the updates of W `on_batch_end`. But that would be relatively elaborate to be honest. Still, its good that you found that out, at least we know we should expect faster results with SGD, right? Did you try with momentum?
